File: 0357-count-numbers-with-unique-digits/0357-count-numbers-with-unique-digits.js
/**
 * @param {number} n
 * @return {number}
 */
var countNumbersWithUniqueDigits = function (n) {
    const dfs = function (length) {
        if (length == n) {
            return 0;
        }

        let count = 0;

        for (let digit = 0; digit < 10; digit++) {
            if (length === 0 && digit === 0) continue;
            if (used[digit]) continue;

            used[digit] = true;
            count += 1;
            count += dfs(length + 1);
            used[digit] = false;
        }

        return count;
    }

    const used = new Array(10).fill(false);
    return 1 + dfs(0);
};

File: 0357-count-numbers-with-unique-digits/0357-count-numbers-with-unique-digits.py
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        def dfs(length):
            if length == n:
                return 0

            count = 0

            for digit in range(10):
                if length == 0 and digit == 0:
                    continue

                if used[digit]:
                    continue

                used[digit] = True
                count += 1

                count += dfs(length + 1)
                used[digit] = False

            return count

        used = [False] * 10
        return 1 + dfs(0)
